### PR TITLE
Spyglass Prow implementation, config, and deployment

### PR DESCRIPTION
**Pull request by krzyzacy**
https://github.com/kubernetes/test-infra/pull/8922 rebased against current master

In this PR:
- Spyglass viewers in prow
- Spyglass config
- Docs
- Deployment file changes

/area prow
/assign @BenTheElder @fejta @cjwagner @paulangton @stevekuznetsov

### CODE REVIEW

**Comment by paulangton** [comment]
CLA should be all set

File: prow/ANNOUNCEMENTS.md [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # Announcements
 
-New features added to each components:
+New features added to each component:
+ - *July 26, 2018* `spyglass` is a pluggable viewing framework for artifacts
+   produced by prowjobs. See a demo [here](prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-performance/121)!
```
Comment (stevekuznetsov) [inline prow/ANNOUNCEMENTS.md]: This 404's

File: prow/cmd/deck/main.go [hunk 2]
```diff
@@ -408,6 +431,171 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
...
+	}
+
+	viewerCache := map[string][]string{}
+	viewersRegistry := ca.Config().Deck.Spyglass.Viewers
+
+	for re, viewerNames := range viewersRegistry {
+		matches := []string{}
+		r, err := regexp.Compile(re)
+		if err != nil {
+			logrus.WithError(err).Error("Regexp failed to compile.")
```
Comment (stevekuznetsov) [inline prow/cmd/deck/main.go]: Can we configure this at config-load time?

File: prow/cmd/deck/main.go [hunk 3]
```diff
@@ -408,6 +431,171 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
...
+		start := time.Now()
+		src := r.URL.Query().Get("src")
+		if src == "" {
+			http.Error(w, "missing src query parameter", http.StatusBadRequest)
+			return
+		}
+
+		fmt.Fprint(w, renderSpyglass(sg, ca, src))
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return view page: ", elapsed)
```
Comment (stevekuznetsov) [inline prow/cmd/deck/main.go]: use a log field or (better) prometheus metric

File: prow/cmd/deck/main.go [hunk 4]
```diff
@@ -408,6 +431,171 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
...
+	}
+	vTmpl := ViewsTemplate{
+		Views:       lenses,
+		Source:      src,
+		ViewerCache: viewerCache,
+	}
+	t := template.Must(template.ParseFiles("/static/template/spyglass-template.html"))
+	tErr := t.Execute(&viewBuf, vTmpl)
+	if tErr != nil {
+		logrus.WithError(tErr).Error("Error rendering template.")
```
Comment (stevekuznetsov) [inline prow/cmd/deck/main.go]: seems like we should return something special in this case

File: prow/cmd/deck/main.go [hunk 5]
```diff
@@ -408,6 +431,171 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
...
+			return
+		} else {
+			pd, err := json.Marshal(lens)
+			if err != nil {
+				logrus.WithError(err).Error("Error marshaling payload.")
+				pd = []byte("{}")
+			}
+		}
+
+		fmt.Fprint(w, string(pd))
```
Comment (stevekuznetsov) [inline prow/cmd/deck/main.go]: responding with 200 always seems wrong here

File: prow/cmd/deck/static/script.js [hunk 6]
```diff
@@ -533,8 +533,11 @@ function redraw(fz) {
         } else {
             r.appendChild(createLinkCell(build.job, build.url, ""));
         }
+
         r.appendChild(createTimeCell(i, parseInt(build.started)));
         r.appendChild(createTextCell(build.duration));
+        // TODO: Uncomment this when ready to deploy Spyglass to the main Deck UI
```
Comment (stevekuznetsov) [inline prow/cmd/deck/static/script.js]: Don't check this in

File: prow/config/config_test.go [hunk 7]
```diff
@@ -29,6 +29,85 @@ import (
 	"k8s.io/test-infra/prow/kube"
 )
 
+func TestSpyglassConfig(t *testing.T) {
+	testCases := []struct {
+		name              string
+		spyglassConfig    string
+		expectedViewers   map[string][]string
+		expectedSizeLimit int64
+		expectError       bool
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: need tests with this true

File: prow/spyglass/README.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
+[![GoDoc Widget]][GoDoc]
+
+# Spyglass
+A spyglass is an lensed monocular maritime instrument used to see things that may have been
```
Comment (stevekuznetsov) [inline prow/spyglass/README.md]: I think my eyes just rolled out of my head ;)

File: prow/spyglass/viewers/buildlog/buildlog_viewer.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+	newLogViewData := LogViewData{
+		Operation: "",
+	}
+	newArtifactView := LogArtifactView{
+		ArtifactName: artifact.JobPath(),
+		ArtifactLink: artifact.CanonicalLink(),
+		ViewName:     name,
+	}
+	switch currentViewData.Operation {
+	case "all":
```
Comment (stevekuznetsov) [inline prow/spyglass/viewers/buildlog/buildlog_viewer.go]: would benefit from

```go
type ViewOperation string
```

?

File: prow/spyglass/viewers/buildlog/buildlog_viewer.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+	}
+	return logLines
+}
+
+// LogViewTemplate executes the log viewer template ready for rendering
+func LogViewTemplate(buildLogsView BuildLogsView) string {
+	var buf bytes.Buffer
+	t := template.Must(template.New(fmt.Sprintf("%sTemplate", name)).Parse(tmplt))
+	if err := t.Execute(&buf, buildLogsView); err != nil {
+		logrus.WithError(err).Error("Error executing template.")
```
Comment (stevekuznetsov) [inline prow/spyglass/viewers/buildlog/buildlog_viewer.go]: Do we get some corrupt data in the buffer here? Should we return some 500-style text?

File: prow/spyglass/viewers/buildlog/buildlog_viewer.go [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+	return logLines
+}
+
+// logLinesAll reads all of an artifact and splits it into lines.
+func logLinesAll(artifact viewers.Artifact) []string {
+	read, err := artifact.ReadAll()
+	if err != nil {
+		if err == viewers.ErrFileTooLarge {
+			logrus.WithError(err).Error("Artifact too large to read all.")
+			return []string{}
```
Comment (stevekuznetsov) [inline prow/spyglass/viewers/buildlog/buildlog_viewer.go]: Seems like we need a mechanism to pass errors up, this is weird

File: prow/spyglass/viewers/junit/junit_template.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package junit
+
+const tmplt = `
```
Comment (stevekuznetsov) [inline prow/spyglass/viewers/junit/junit_template.go]: I continue to think that development on these will be a real chore. There is nothing nice about the UX of embedding one language in another. No syntax checks, no highlighting, etc. It's going to be really hard to review... can we build this more programattically, or add some functionality to check these for validity so local iteration is useful?

**Comment by fejta** [comment]
pass tests plesae!

File: prow/ANNOUNCEMENTS.md (hunk 1 above)
Comment (paulangton) [reply prow/ANNOUNCEMENTS.md]: it 404's because spyglass isnt deployed yet

**Comment by krzyzacy** [comment]
/hold
I'll get back to this later in the week and address some feedback first

**Comment by paulangton** [comment]
@krzyzacy I dont have push permissions to this branch, I'm going to pull this branch into mine and create a separate PR for it

File: prow/cmd/deck/main.go (hunk 4 above)
Comment (paulangton) [reply prow/cmd/deck/main.go]: yes this seems like a good place for a 500

File: prow/cmd/deck/main.go (hunk 3 above)
Comment (paulangton) [reply prow/cmd/deck/main.go]: metrics will be exported in #8956

File: prow/cmd/deck/main.go (hunk 2 above)
Comment (paulangton) [reply prow/cmd/deck/main.go]: Definitely can, wasnt sure if we wanted to block loading config on an invalid regex. Anyone else have opinions on this?

File: prow/cmd/deck/main.go (hunk 5 above)
Comment (paulangton) [reply prow/cmd/deck/main.go]: Agree, writing a 500 to the individual view when a refresh request fails

File: prow/spyglass/viewers/junit/junit_template.go (hunk 12 above)
Comment (paulangton) [reply prow/spyglass/viewers/junit/junit_template.go]: This is how I chose to develop my version of the view, I agree that this isnt ideal for syntax highlighting, but the viewer implementations are not the focus here. Follow-up PRs are welcome to change the standard viewer development workflow to something more friendly

File: prow/spyglass/viewers/buildlog/buildlog_viewer.go (hunk 11 above)
Comment (paulangton) [reply prow/spyglass/viewers/buildlog/buildlog_viewer.go]: Pass errors up? If the viewers find they are unable to deal with a certain artifact they can just return a div with the error no?

File: prow/spyglass/viewers/buildlog/buildlog_viewer.go (hunk 10 above)
Comment (paulangton) [reply prow/spyglass/viewers/buildlog/buildlog_viewer.go]: Yes we can do that, addressing all viewer-implementation feedback in viewer-specific PRs

File: prow/cmd/deck/main.go (hunk 2 above)
Comment (paulangton) [reply prow/cmd/deck/main.go]: @cjwagner @krzyzacy

File: prow/spyglass/viewers/buildlog/buildlog_viewer.go (hunk 9 above)
Comment (paulangton) [reply prow/spyglass/viewers/buildlog/buildlog_viewer.go]: Without proper string enums what does this actually gain us?

**Comment by fejta** [comment]
Ack

**Comment by fejta** [comment]
Close?

**Pull request closed without merging**